### PR TITLE
[Intel MKL] fix bias cache accuracy issue.

### PR DESCRIPTION
**Pull request by guizili0**
This PR fix the bias cache optimization accuracy reverted in https://github.com/tensorflow/tensorflow/pull/32922

### CODE REVIEW

File: tensorflow/core/kernels/mkl_conv_ops.cc [hunk 1]
```diff
@@ -1590,19 +1588,17 @@ class MklQuantizedConv2DOp
...
-        }
+      size_t depth = min_filter_vector.NumElements();
+      std::vector<float> scales(depth);
+      for (size_t i = 0; i < depth; ++i) {
+        scales[i] =
+            int_const_scale_limit /
+            (std::max(std::abs(max_input), std::abs(min_input)) *
+             std::max(std::abs(max_filter[i]), std::abs(min_filter[i])));
+      }
+      if (!is_bias_const_ || IsBiasCacheEmpty(context) ||
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_conv_ops.cc]: To make the code easier to understand, could you please restructure it this way?

- Check if bias is enabled at the beginning of the function (`GetBiasHandle`) and return early if so. 
```c++
if (!bias_enabled) return nullptr;
```
- Take the rest of the code out of the current `if (bias_enabled)`.
- Return the `GetCachedBias` case first, checking if scales are valid on the go (not storing in vector). Something like this:
```c++
// If scales_ exists, checks if the current scales are the same.
bool scales_are_valid = (depth == scales_.size());
const float intermediate_scale_limit = 
    int_const_scale_limit / (std::max(std::abs(max_input), std::abs(min_input));
for (size_t i = 0; scales_are_valid && i < depth; ++i) {
  float multiplier = std::max(std::abs(max_filter[i]), std::abs(min_filter[i]));
  if (std::abs(intermediate_scale_limit - scales_[i] * multiplier) > some_epsilon) {
    scales_are_valid = false;
  }
}
// The bias has been cached and is safe to use.
if (is_bias_const_ && !IsBiasCacheEmpty(context) && scales_are_valid) {
  return GetCachedBias(context);
}
// The rest.
```

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you for your PR! I think the code should be restructured a bit to convey the logic a little clearer.

File: tensorflow/core/kernels/mkl_conv_ops.cc (hunk 1 above)
Comment (guizili0) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: @penpornk As I did not know the scales are valid until I make sure all the scales are some to the original one. So I just can stop to compare the two float if the scales already invalid. I also need to compute all the scales and set to the new one.
Do you have any comments here, thanks.

**Comment by guizili0** [comment]
@penpornk thanks for you comments, I have modified my code based on it, please help to check, thanks.

File: tensorflow/core/kernels/mkl_conv_ops.cc [hunk 2]
```diff
@@ -1576,65 +1576,66 @@ class MklQuantizedConv2DOp
     const float* max_filter = max_filter_vector.flat<float>().data();
 
     std::vector<mkldnn::primitive> net;
-    if (bias_enabled) {
-      if (std::is_same<Tbias, qint32>::value) {
-        return static_cast<Tbias*>(
-            const_cast<Tbias*>(bias_tensor.flat<Tbias>().data()));
-      }
+    if (!bias_enabled) {
```
Comment (penpornk) [inline tensorflow/core/kernels/mkl_conv_ops.cc]: This and the next `if` can be moved to the beginning of the function (line 1566, before `int bias_index_offset`).

**Review by penpornk — CHANGES_REQUESTED** [review]
Thank you! One minor change please.

File: tensorflow/core/kernels/mkl_conv_ops.cc (hunk 2 above)
Comment (guizili0) [reply tensorflow/core/kernels/mkl_conv_ops.cc]: @penpornk Modified by you request. Please check, thanks.

**Review by penpornk — APPROVED** [review]
Thank you for your patience!

**Pull request merged**